pkg/server/api/agent/v1/service.go

```diff
@@ -471,3 +618,206 @@ func (s *Service) getSelectorsFromAgentID(ctx context.Context, agentID string) (
 
 	return api.NodeSelectorsToProto(resp.Selectors)
 }
+
+func validateAttestAgentParams(params *agent.AttestAgentRequest_Params) error {
+	if params == nil {
+		return errors.New("missing params")
+	}
+	if params.Data == nil {
+		return errors.New("missing attestation data")
+	}
+	if params.Params == nil {
+		return errors.New("missing X509-SVID parameters")
+	}
+	if len(params.Params.Csr) == 0 {
+		return errors.New("missing CSR")
+	}
+	if params.Data.Type == "" {
+		return errors.New("missing attestation data type")
+	}
+	if params.Data.Payload == "" {
+		return errors.New("missing attestation data payload")
+	}
+	return nil
+}
+
+func (s *Service) isAttested(node *common.AttestedNode, baseSpiffeID string) bool {
+	if node != nil && node.SpiffeId == baseSpiffeID {
+		return true
+	}
+	return false
+}
+
+func (s *Service) isBanned(node *common.AttestedNode) bool {
+	if node != nil && nodeutil.IsAgentBanned(node) {
+		return true
+	}
+	return false
+}
+
+func getAttestAgentResponse(spiffeID string, certificates []*x509.Certificate) (*agent.AttestAgentResponse, error) {
+	ID, err := spiffeid.FromString(spiffeID)
+	if err != nil {
+		return nil, err
+	}
+	protoID := api.ProtoFromID(ID)
+
+	svid := &types.X509SVID{
+		Id:        protoID,
+		CertChain: x509util.RawCertsFromCertificates(certificates),
+		ExpiresAt: certificates[0].NotAfter.Unix(),
+	}
+
+	return &agent.AttestAgentResponse{
+		Step: &agent.AttestAgentResponse_Result_{
+			Result: &agent.AttestAgentResponse_Result{
+				Svid: svid,
+			},
+		},
+	}, nil
+}
+
+func (s *Service) attestJoinToken(ctx context.Context, token string) (*nodeattestor.AttestResponse, error) {
+	log := rpccontext.Logger(ctx).WithField(telemetry.NodeAttestorType, "join_token")
+
+	resp, err := s.ds.FetchJoinToken(ctx, &datastore.FetchJoinTokenRequest{
+		Token: token,
+	})
+	if err != nil {
+		log.WithError(err).Error("Failed to fetch join token")
+		return nil, status.Error(codes.Internal, "failed to fetch join token")
+	}
+
+	if resp.JoinToken == nil {
+		log.Error("Join token does not exist or has already been used")
+		return nil, status.Error(codes.InvalidArgument, "join token does not exist or has already been used")
+	}
+
+	if resp.JoinToken.Token == "" {
+		log.Error("Invalid join token")
+		return nil, status.Error(codes.InvalidArgument, "invalid join token")
+	}
```

Comment:
Under what conditions will Token == ""? I'm not sure this check is necessary. If the JoinToken message ended up having an empty token value (presumably the fetch call would fail) it doesn't seem to impact the rest of the logic. If it is important, it seems to represent an invalid internal state and should be `codes.Internal`.

Reply:
I wasn't sure under what conditions that situation would happen. I saw that validation in the old implementation, found it extremely defensive, but decided to leave it in case I was missing something.